File: python/seed_format.py
"""Binary .seedmodel format — cross-language seed model serialization.

Format:
    SDML                          <- 4-byte magic
    version: u8                   <- format version (1)
    seed_id: u8                   <- seed type ID (0-255)
    max_order: u8                 <- max PPM order stored
    name_len: u8                  <- length of name string
    name: [u8; name_len]          <- UTF-8 name (e.g. "english")

    For each order 0..max_order:
      num_contexts: u32 (LE)
      For each context:
        context_bytes: [u8; order]
        num_entries: u16 (LE)     <- number of non-zero symbols (sparse)
        For each entry:
          symbol: u8
          count: u16 (LE)
"""

import struct

MAGIC = b"SDML"
FORMAT_VERSION = 1
NUM_SYMBOLS = 256


def write_seed(path, seed_id, name, max_order, counts):
    """Serialize seed counts to a .seedmodel file.

    Args:
        path: Output file path.
        seed_id: Seed type ID (0-255).
        name: Human-readable name (e.g. "english").
        max_order: Max PPM order stored.
        counts: {order: {context_bytes: [256 counts]}}
    """
    name_bytes = name.encode("utf-8")
    buf = bytearray()

    # Header
    buf.extend(MAGIC)
    buf.append(FORMAT_VERSION)
    buf.append(seed_id)
    buf.append(max_order)
    buf.append(len(name_bytes))
    buf.extend(name_bytes)

    # Per-order data
    for order in range(max_order + 1):
        order_counts = counts.get(order, {})
        buf.extend(struct.pack("<I", len(order_counts)))
        for ctx, syms in order_counts.items():
            assert len(ctx) == order, f"context length {len(ctx)} != order {order}"
            buf.extend(ctx)
            # Sparse: only non-zero entries
            entries = [(s, c) for s, c in enumerate(syms) if c > 0]
            buf.extend(struct.pack("<H", len(entries)))
            for symbol, count in entries:
                buf.append(symbol)
                buf.extend(struct.pack("<H", count))

    with open(path, "wb") as f:
        f.write(buf)


def read_seed(path):
    """Deserialize a .seedmodel file.

    Returns:
        (seed_id, name, max_order, counts)
        counts: {order: {context_bytes: [256 counts]}}
    """
    with open(path, "rb") as f:
        data = f.read()

    if data[:4] != MAGIC:
        raise ValueError("Not a .seedmodel file (bad magic)")

    off = 4
    version = data[off]; off += 1
    if version != FORMAT_VERSION:
        raise ValueError(f"Unsupported seedmodel version {version}")

    seed_id = data[off]; off += 1
    max_order = data[off]; off += 1
    name_len = data[off]; off += 1
    name = data[off:off + name_len].decode("utf-8"); off += name_len

    counts = {}
    for order in range(max_order + 1):
        num_contexts = struct.unpack_from("<I", data, off)[0]; off += 4
        order_counts = {}
        for _ in range(num_contexts):
            ctx = bytes(data[off:off + order]); off += order
            num_entries = struct.unpack_from("<H", data, off)[0]; off += 2
            syms = [0] * NUM_SYMBOLS
            for _ in range(num_entries):
                symbol = data[off]; off += 1
                count = struct.unpack_from("<H", data, off)[0]; off += 2
                syms[symbol] = count
            order_counts[ctx] = syms
        counts[order] = order_counts

    return seed_id, name, max_order, counts

File: src/seed_format.rs
/// Binary .seedmodel format reader.
///
/// Bit-identical to the Python reference implementation in python/seed_format.py.

use std::collections::HashMap;
use std::fs;
use std::path::Path;

const MAGIC: &[u8; 4] = b"SDML";
const FORMAT_VERSION: u8 = 1;
pub const NUM_SYMBOLS: usize = 256;

/// Seed counts: order -> (context_bytes -> [256 counts])
pub type SeedCounts = HashMap<u8, HashMap<Vec<u8>, Vec<u32>>>;

#[derive(Debug)]
pub struct SeedModel {
    pub seed_id: u8,
    pub name: String,
    pub max_order: u8,
    pub counts: SeedCounts,
}

pub fn read_seed(path: &Path) -> Result<SeedModel, String> {
    let data = fs::read(path).map_err(|e| format!("Failed to read {}: {}", path.display(), e))?;
    parse_seed(&data)
}

pub fn parse_seed(data: &[u8]) -> Result<SeedModel, String> {
    if data.len() < 8 || &data[..4] != MAGIC {
        return Err("Not a .seedmodel file (bad magic)".into());
    }

    let mut off = 4;
    let version = data[off];
    off += 1;
    if version != FORMAT_VERSION {
        return Err(format!("Unsupported seedmodel version {}", version));
    }

    let seed_id = data[off];
    off += 1;
    let max_order = data[off];
    off += 1;
    let name_len = data[off] as usize;
    off += 1;

    if off + name_len > data.len() {
        return Err("Truncated name".into());
    }
    let name = String::from_utf8(data[off..off + name_len].to_vec())
        .map_err(|_| "Invalid UTF-8 in name")?;
    off += name_len;

    let mut counts: SeedCounts = HashMap::new();

    for order in 0..=max_order {
        if off + 4 > data.len() {
            return Err("Truncated data".into());
        }
        let num_contexts = u32::from_le_bytes([data[off], data[off + 1], data[off + 2], data[off + 3]]) as usize;
        off += 4;

        let mut order_counts: HashMap<Vec<u8>, Vec<u32>> = HashMap::new();
        let order_usize = order as usize;

        for _ in 0..num_contexts {
            if off + order_usize > data.len() {
                return Err("Truncated context".into());
            }
            let ctx = data[off..off + order_usize].to_vec();
            off += order_usize;

            if off + 2 > data.len() {
                return Err("Truncated entries header".into());
            }
            let num_entries = u16::from_le_bytes([data[off], data[off + 1]]) as usize;
            off += 2;

            let mut syms = vec![0u32; NUM_SYMBOLS];
            for _ in 0..num_entries {
                if off + 3 > data.len() {
                    return Err("Truncated entry".into());
                }
                let symbol = data[off] as usize;
                off += 1;
                let count = u16::from_le_bytes([data[off], data[off + 1]]) as u32;
                off += 2;
                syms[symbol] = count;
            }
            order_counts.insert(ctx, syms);
        }
        counts.insert(order, order_counts);
    }

    Ok(SeedModel {
        seed_id,
        name,
        max_order,
        counts,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    fn seeds_dir() -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("seeds")
    }

    #[test]
    fn test_read_null() {
        let path = seeds_dir().join("null.seedmodel");
        let seed = read_seed(&path).expect("Failed to read null.seedmodel");
        assert_eq!(seed.seed_id, 0);
        assert_eq!(seed.name, "null");
        assert_eq!(seed.max_order, 4);
    }

    #[test]
    fn test_read_english() {
        let path = seeds_dir().join("english.seedmodel");
        let seed = read_seed(&path).expect("Failed to read english.seedmodel");
        assert_eq!(seed.seed_id, 1);
        assert_eq!(seed.name, "english");
        assert_eq!(seed.max_order, 4);
        // Should have contexts at order 0
        let order0 = seed.counts.get(&0).unwrap();
        assert!(!order0.is_empty());
    }

    #[test]
    fn test_bad_magic() {
        let data = b"BAAD\x01\x00\x04\x04null";
        let result = parse_seed(data);
        assert!(result.is_err());
        assert!(result.unwrap_err().contains("bad magic"));
    }
}
